File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        for(int i = 0 ; i < nums.length; i++){
            if(i != 0 && nums[i-1] == nums[i]) continue;
            int target = 0 - nums[i];
            int l = i + 1 , r = nums.length - 1;
            while(l < r){
                int x = nums[l] + nums[r];
                if(x < target){
                    l++;
                }else if(x > target){
                    r--;
                }else{
                    List<Integer> a = new ArrayList<>();
                    a.add(nums[i]);
                    a.add(nums[l]);
                    a.add(nums[r]);
                    ans.add(a);
                    while(l < nums.length - 1 && nums[l] == nums[l+1]) l++;
                    while(r > i && nums[r-1] == nums[r]) r--;
                    r--; l++;
                }
            }
        }
        return ans;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums = sorted(nums)
        for i, num in enumerate(nums):
            if i > 0 and num == nums[i - 1]:
                continue
            target = -num
            left, right = i + 1, len(nums) - 1
            # print('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')
            while left < right:
                # print(num, nums[left], nums[right])
                if left > i + 1 and nums[left] == nums[left - 1]:
                    left += 1
                elif right < len(nums) - 1 and nums[right] == nums[right + 1]:
                    right -= 1
                else:
                    val = nums[left] + nums[right]
                    if val > target:
                        right -= 1
                    elif val < target:
                        left += 1
                    else:
                        ans.append([num, nums[left], nums[right]])
                        left += 1
                        right -= 1
        return ans
                    
        